src/main/scala/is/hail/stats/SkatModel.scala

```diff
@@ -0,0 +1,211 @@
+    *    for the Skat stat.
+    * Input:
+    *   Grammian: (m x m) DenseMatrix
+    *     the matrix with the same eigenvalues of P0^{1/2}KP0^{1/2} from the original
+    *     paper.
+    * Returns:
+    *   ((qstat, pval), fault)      SkatStat
+    *     SKAT statistic and p-value computed by Davies Algorithm returned as a instance
+    *     of the SkatStat class. With the statistics a integer indicating any issues
+    *     running Davies Algorithm is also returned.
+    */
+  def computePVal(Grammian: DenseMatrix[Double]): SkatStat = {
+    val allEvals = eigSymD.justEigenvalues(Grammian).toArray
+
+    //compute threshold for eigenvalues
+    var allEvalsSum = 0.0
+    var i = 0
+    while (i < allEvals.length) {
+      allEvalsSum += allEvals(i)
+      i += 1
+    }
+    val threshold = 1e-5 * allEvalsSum / allEvals.length
+
+    //Initialize parameters to Davies' algorithm
+    val evals = allEvals.filter(_ > threshold)
+    val terms = evals.length
+
+    val noncentrality = Array.ofDim[Double](terms)
+    val dof = Array.fill[Int](terms)(1)
+    val trace = Array.fill[Double](7)(0.0)
+    val fault = new IntByReference()
+    val s = 0.0
+    val accuracy = 1e-6
+    val iterations = 10000
+    val x = qfWrapper(evals, noncentrality, dof, terms, s, skatStat, iterations, accuracy, trace, fault)
+
+    SkatStat(skatStat, 1 - x, fault.getValue)
+
+  }
+
+  /** computeLinearSkatStats
+    *   Computes the p value of the variance-component score statistic as
+    *   defined by SKAT.
+    * Input:
+    *   GwGrammian              (m x m) DenseMatrix
+    *     The Gram matrix of the genotype matrix for n samples and m genetic
+    *     variants with each column scaled by the corresponding variant weight.
+    *   QtGwGrammian            (m x m) DenseMatrix
+    *     The Gram matrix of the projection of the weighted genotype matrix Gw
+    *     by Q.t, where Q comes from the reduced QR decomposition of the
+    *     covariate data.
+    * Returns:
+    *   SkatStat (see computePVal)
+    */
+  def computeLinearSkatStats(GwGrammian: DenseMatrix[Double], QtGwGrammian: DenseMatrix[Double]): SkatStat = {
+
+    val variantGrammian = 0.5 * (GwGrammian - QtGwGrammian)
+    computePVal(variantGrammian)
+  }
+
+  /** LogisticSKATModel
+    *   Computes the p value of the variance-component score statistic as
+    *   defined by SKAT from a logistic fit.
+    * Input:
+    *   X        (n x k) DenseMatrix
+    *     The covariates matrix associated with the regression for n samples and
+    *     k covariates.
+    *   (TODO) mu (n x 1) DenseVector
+    *     The expectation computed from the regression model.
+    *   Gw       (n x m) DenseMatrix
+    *     The weighted genotype matrix corresponding to the variants in a
+    *     particular gene.
+    * Returns:
+    *   SkatStat (see computePVal)
+    */
+  def computeLogisticSkatStat(X: DenseMatrix[Double], mu: DenseVector[Double],
+    Gw: DenseMatrix[Double]): Unit = {
```

Comment:
that's just there because the function was still being built. It is replaced with SkatStat on the logisticSkat branch